tests/test_cliCmdQueryDatasets.py

```diff
@@ -0,0 +1,260 @@
+        datasetTypeName = "alt_test_metric_comp"
+        storageClass = self.storageClassFactory.getStorageClass("StructuredCompositeReadComp")
+        datasetType = self._addDatasetType(datasetTypeName, dimensions, storageClass, self.butler.registry)
+        self.butler.registry.insertDimensionData(
+            "visit", {"instrument": "DummyCamComp", "id": 425,
+                      "name": "fourtwentyffive", "physical_filter": "d-r",
+                      "visit_system": 1})
+        metric = self._makeExampleMetrics()
+        dataId = {"instrument": "DummyCamComp", "visit": 425}
+        ref = DatasetRef(datasetType, dataId, id=None)
+        self.butler.put(metric, ref)
+
+        # verify the new dataset type increases the number of tables found:
+        tables = self._queryDatasets(repo=self.butlerConfigFile)
+        self.assertEqual(len(tables), 6)
+        types = [table["type"][0] for table in tables]
+        self.assertEqual(set(types), set(["test_metric_comp.data",
+                                          "test_metric_comp.output",
+                                          "test_metric_comp.summary",
+                                          "alt_test_metric_comp.data",
+                                          "alt_test_metric_comp.output",
+                                          "alt_test_metric_comp.summary"]))
+
+        # verify the new dataset type increases the number of tables found:
+        tables = self._queryDatasets(repo=self.butlerConfigFile, glob=("alt_test_*", ))
+        self.assertEqual(len(tables), 3)
+        types = [table["type"][0] for table in tables]
+        self.assertEqual(set(types), set(["alt_test_metric_comp.data",
+                                          "alt_test_metric_comp.output",
+                                          "alt_test_metric_comp.summary"]))
+
+    @staticmethod
+    def _makeUri(*segments):
+        return urlunparse(urlparse(os.path.join(*segments), scheme="file"))
+
+    def testFindFirstAndCollections(self):
+        """Test the find-first option, and the collections option, since it
+        is required for find-first."""
+
+        # Create a new butler with a new run, and add a dataset to shadow an
+        # existing dataset.
+        self.butler = Butler(self.root, run="foo")
+        metric = self._makeExampleMetrics()
+        dataId = {"instrument": "DummyCamComp", "visit": 424}
+        ref = DatasetRef(self.datasetType, dataId, id=None)
+        self.butler.put(metric, ref)
+
+        # Verify that without find-first, duplicate datasets are returned
+        tables = self._queryDatasets(repo=self.butlerConfigFile,
+                                     collections=["foo", "ingest/run"],
+                                     show_uri=True)
+        self.maxDiff = None
+        print(self.root)
+        self.assertListEqual(tables[0]["URI"].tolist(), [
+            self._makeUri(
```

Comment:
The Butler datastore root is a ButlerURI so for what you are doing something like `butler.datastore.root.join("ingest/run/test_blah/")` should work for you without having to use urlunparse.